worker/caasrbacmapper/mapper.go

```diff
@@ -0,0 +1,167 @@
+	Mapper
+	worker.Worker
+}
+
+type DefaultMapper struct {
+	catacomb     catacomb.Catacomb
+	lock         *sync.RWMutex
+	logger       Logger
+	saInformer   core.ServiceAccountInformer
+	saNameUIDMap map[string]types.UID
+	saUIDAppMap  map[types.UID]string
+	workQueue    workqueue.RateLimitingInterface
+}
+
+func (d *DefaultMapper) AppNameForServiceAccount(id types.UID) (string, error) {
+	d.lock.RLock()
+	defer d.lock.RUnlock()
+	appName, found := d.saUIDAppMap[id]
+	if !found {
+		return "", errors.NotFoundf("no service account for app found with id %v", id)
+	}
+	return appName, nil
+}
+
+func (d *DefaultMapper) enqueueServiceAccount(obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		d.logger.Errorf("failed enqueuing service account: %v", err)
+		return
+	}
+	d.workQueue.Add(key)
+}
+
+func (d *DefaultMapper) Kill() {
+	d.catacomb.Kill(nil)
+}
+
+func (d *DefaultMapper) loop() error {
+	defer d.workQueue.ShutDown()
+	d.saInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    d.enqueueServiceAccount,
+		DeleteFunc: d.enqueueServiceAccount,
+		UpdateFunc: func(_, newObj interface{}) {
+			d.enqueueServiceAccount(newObj)
+		},
+	})
+
+	go d.saInformer.Informer().Run(d.catacomb.Dying())
+	go wait.Until(func() {
+		for d.processNextQueueItem() {
+		}
+	}, time.Second, d.catacomb.Dying())
+
+	select {
+	case <-d.catacomb.Dying():
+	}
+
+	return nil
+}
+
+func NewMapper(logger Logger, informer core.ServiceAccountInformer) (*DefaultMapper, error) {
+	dm := &DefaultMapper{
+		lock:         new(sync.RWMutex),
+		logger:       logger,
+		saInformer:   informer,
+		saNameUIDMap: map[string]types.UID{},
+		saUIDAppMap:  map[types.UID]string{},
+		workQueue:    workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+	}
+
+	if err := catacomb.Invoke(catacomb.Plan{
+		Site: &dm.catacomb,
+		Work: dm.loop,
+	}); err != nil {
+		return dm, errors.Trace(err)
+	}
+	return dm, nil
+}
+
+func (d *DefaultMapper) processNextQueueItem() bool {
+	obj, shutdown := d.workQueue.Get()
+	if shutdown {
+		return false
+	}
+
+	defer d.workQueue.Done(obj)
+	key, ok := obj.(string)
+	if !ok {
+		d.logger.Errorf("failed converting service account queue item to string")
+		return false
+	}
+
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		d.logger.Errorf("failed spliting key into namespace and name for service account queue: %v", err)
+		return false
+	}
+
+	sa, err := d.saInformer.Lister().ServiceAccounts(namespace).Get(name)
+	if k8serrors.IsNotFound(err) {
+		d.lock.Lock()
+		defer d.lock.Unlock()
+		uid, exists := d.saNameUIDMap[key]
+		if !exists {
+			return true
+		}
+		delete(d.saUIDAppMap, uid)
+		delete(d.saNameUIDMap, name)
+		return true
+	}
+	if err != nil {
+		d.logger.Errorf("failed fetching service account for %s/%s", namespace, name)
+		return false
+	}
+
+	appName, err := provider.AppNameForServiceAccount(sa)
+	if errors.IsNotFound(err) {
```

Comment:
d.workqueue.Forget(obj) here as well?
because we don't care non-juju created resources

Reply:
Ah nice spot, Done & Done